### PR TITLE
Modify CTS for reduced feature set compatibility, add tests

### LINKED ISSUE #365: Add support for reduced feature sets.
A number of SYCL-CTS tests require "full" feature set. For instance, multi_ptr test requires implementation to support generic address space and id test requires implementation to support unnamed SYCL kernel function feature.

### PR DESCRIPTION
**Pull request by nolmoonen**
Closes #365 and implements test plan #473.
- Every kernel launch is supplied with a kernel name (`single_task`, `parallel_for`, `parallel_for_work_group`). This is done in commit "add kernel names" (with the exception of some fixes in the last commit).
- The test described in #473 is added in `tests/full_feature_set`.
- The `handler` test is extended to also launch nameless kernels when `SYCL_CTS_ENABLE_FEATURE_SET_FULL` is `ON`.
- The `multi_ptr` tests that rely on the generic address space being the default are guarded with `#if SYCL_CTS_ENABLE_FEATURE_SET_FULL`.

A topic of debate may be whether `SYCL_CTS_ENABLE_FEATURE_SET_FULL` should be enabled by default. Right now, that is the case and the CI configuration script for ComputeCpp turns it off so that the CI may still pass.

**The CI stage for ComputeCpp still fails because that Docker image is not automatically rebuilt, hence not taking the changes to the configure script into account.**

### CODE REVIEW

File: tests/id/id.cpp [hunk 1]
```diff
@@ -17,6 +29,39 @@
 
 using namespace sycl_cts;
 
+template <typename T, std::size_t Line, int Dimensions = 0>
+class kernel_name;
+
+/** Define a unique kernel name for \p DEVICE_EVAL. */
+#define EVAL(expr) DEVICE_EVAL(expr, kernel_name<decltype(expr), __LINE__>)
```
Comment (psalz) [inline tests/id/id.cpp:36]: Nice that you found a way of making `DEVICE_EVAL` work for the reduced feature set!

**Review by ProGTX — APPROVED** [review]

**Comment by ProGTX** [comment]
> A topic of debate may be whether SYCL_CTS_ENABLE_FEATURE_SET_FULL should be enabled by default. Right now, that is the case and the CI configuration script for ComputeCpp turns it off so that the CI may still pass.

I think that makes sense. The important part is that the CMake options, including this one, are included in the conformance report.

> The CI stage for ComputeCpp still fails because that Docker image is not automatically rebuilt, hence not taking the changes to the configure script into account.

So we should make the changes in the Docker file on our end, right?

**Comment by nolmoonen** [comment]
> I think that makes sense. The important part is that the CMake options, including this one, are included in the conformance report.

I was not aware of the script for the generation of a conformance report, I've added the option to it as well.

> So we should make the changes in the Docker file on our end, right?

As far as I understand it, yes. Note that specifying this CMake option without the changes in this PR will only throw a CMake warning, so it does not require this PR to be merged first.

**Comment by fraggamuffin** [comment]
needs an option in CTS since default is full feature

**Comment by bader** [comment]
> Force push to resolve merge conflicts.

Unfortunately, there are still merge conflicts and code doesn't compile with computecpp.

**Review by bader — APPROVED** [review]

File: docs/README.adoc [hunk 2]
```diff
@@ -417,3 +417,6 @@ While legacy test cases are still mapped to Catch2 under the hood, they require
 Although technically not required, usually only one class extending `test_base` is defined per file.
 
 IMPORTANT: Always write new-style test cases.
+
+=== Reduced and full feature set
+SYCL2020 specifies a full and a reduced feature set. The full feature set includes all features in the core SYCL specification without exceptions, the reduced feature set makes certain features optional. To ensure that reduced feature set implementations can test conformance, the CTS should by default not include features that are in the full feature set, but not in the reduced feature set. The CTS option `SYCL_CTS_ENABLE_FEATURE_SET_FULL` may be set to test conformance for the full feature set, and this macro can be used to specify additional tests for features that are only in the full feature set.
```
[the remark this answers is not in the document]
Comment (nolmoonen) [reply docs/README.adoc]: The purpose of this sentence is to say that the CTS should not include features of the full feature set without a guard. I have updated the wording to that it is clearer.

**Pull request merged**